### PR TITLE
CNKAF-739 - Reassignments endpoint for given topic name and partition id

### PR DESCRIPTION
**Pull request by agg111**
### What
This PR adds endpoint to list on reassignment for a topic name and partition id - 
``` GET /v3/clusters/<clusterId>/topics/{topicName}/partitions/{partitionId}/reassignments```

### Testing
Added unit tests for resources and controller classes.
Integration test to verify request results.
```mvn test```

### References
One pager - https://confluentinc.atlassian.net/wiki/spaces/CNKAF/pages/1276742825/REST+API+for+partition+reassignment
Jira - https://confluentinc.atlassian.net/browse/CNKAF-739

### CODE REVIEW

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import io.confluent.kafkarest.entities.Reassignment;
+import io.confluent.kafkarest.entities.Topic;
+import io.confluent.kafkarest.entities.Partition;
+import io.confluent.kafkarest.entities.Cluster;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+public interface ReassignmentManager {
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java:26]: Should we call this `PartitionReassignmentManager`. Even if the API is `/reassignments`, it's more clear to be explicit here.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  /**
+   * Returns a list of Kafka {@link Reassignment} for all Kafka {@link Topic} reassignments for a
+   * Kafka {@link Cluster} with the given {@code clusterId}.
+   */
+  CompletableFuture<List<Reassignment>> listReassignments(String clusterId);
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java:32]: Likewise, `PartitionReassignment`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.PartitionReassignment;
+import org.apache.kafka.common.TopicPartition;
+
+public class ReassignmentManagerImpl implements ReassignmentManager {
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:35]: package-private and final

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            });
+  }
+
+  @Override
+  public CompletableFuture<Optional<Reassignment>> getReassignment(
+      String clusterId, String topicName, Integer partitionId) {
+    return listReassignments(clusterId)
+        .thenApply(reassignments -> reassignments.stream()
+            .filter(reassignment -> reassignment.getTopicName().equals(topicName)
+                && reassignment.getPartitionId() == partitionId)
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:76]: Super nit: You can do `.filter(...).filter(...)`. I think it looks better than `.filter(... && ...)`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            .findAny());
+  }
+
+  private static Reassignment toReassignment(String clusterId,
+      Entry<TopicPartition, PartitionReassignment> reassignment) {
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:81]: `Entry` by itself is a non-descriptive name. Use the fullname `Map.Entry`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    return Reassignment.create(
+        clusterId,
+        reassignment.getKey().topic(),
+        reassignment.getKey().partition(),
+        reassignment.getValue().replicas(),
+        reassignment.getValue().addingReplicas(),
+        reassignment.getValue().removingReplicas());
+  }
+
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:90]: nit: remove new-line

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  Reassignment() {
+  }
+
+  public abstract String getClusterId();
+
+  public abstract String getTopicName();
+
+  public abstract int getPartitionId();
+
+  public abstract List<Integer> getReplicas();
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java:34]: Use `ImmutableList` for here (keep `List in the `create`).

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java (hunk 4 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java]: Thanks for this one.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java (hunk 3 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java]: Updated.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java (hunk 1 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java:26]: As discussed offline, will leave it as `Reassignments`

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java (hunk 5 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java]: Updated.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java (hunk 7 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java]: Made it `ImmutableList`. Keeping the argument passed to create as` List`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java (hunk 2 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java:32]: Leaving it as `Reassignments` after the offline sync.

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/GetReassignmentAction.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    CompletableFuture<GetReassignmentResponse> response =
+        reassignmentManager.get().getReassignment(clusterId, topicName, partitionId)
+            .thenApply(reassignment -> reassignment.orElseThrow(NotFoundException::new))
+            .thenApply(
+                reassignment -> GetReassignmentResponse.create(toReassignmentData(reassignment)));
+
+    AsyncResponses.asyncResume(asyncResponse, response);
+  }
+
+  private ReassignmentData toReassignmentData(Reassignment reassignment) {
```
Comment (agg111) [inline kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/GetReassignmentAction.java:73]: This function is reduntant to the one in `ListAllReassignmentAction.java`, so thinking if it is okay to have it at one place chaning the scope of it.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java [hunk 9]
```diff
@@ -19,10 +19,6 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 import com.google.auto.value.AutoValue;
 
-/**
- * Response body for {@code GET /v3/clusters/<clusterId>/topics/-/partitions/-/reassignments}
- * requests.
- */
```
Comment (agg111) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java:25]: Removed this comment to maintain consistency with other response classes.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java [hunk 10]
```diff
@@ -19,10 +19,6 @@
...
 @AutoValue
 public abstract class ListAllReassignmentsResponse {
```
Comment (agg111) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java:23]: This class can actually be renamed to just ListReassignmentsResponse. The "all" is valid for request  - `GET /v3/clusters/<clusterId>/topics/-/partitions/-/reassignments`

Comment (rigelbm) [reply kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java:23]: I think you should have three classes:

* `ListAllReassignmentsResponse` for `/v3/clusters/<clusterId>/topics/-/partitions/-/reassignments` (#682)
* `SearchReassignmentsByTopicResponse` for `/v3/clusters/<clusterId>/topics/<topicName>/partitions/-/reassignments` (#687)
* `GetReassignmentsResponse` for `/v3/clusters/<clusterId>/topics/<topicName>/partitions/<partitionId>/reassignments` (this PR)

Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListAllReassignmentsResponse.java:23]: That sounds good, updated #687 with separate response class as `SearchReassignmentsByTopicResponse`.

File: kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/GetReassignmentAction.java (hunk 8 above)
Comment (rigelbm) [reply kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/GetReassignmentAction.java:73]: You can have it public static (receiving CrnFactory and UrlFactory as argument). You you can just copy and paste. Your choice.

**Pull request merged**